clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -229,7 +229,7 @@ public long connectionDelay(Node node, long now) {
      */
     @Override
     public boolean connectionFailed(Node node) {
-        return connectionStates.connectionState(node.idString()).equals(ConnectionState.DISCONNECTED);
+        return connectionStates.isDisconnected(node.idString());
```

Comment:
@ijuma You mentioned in your first comment that "It's weird to say that a connection that has never been attempted has failed", which seems to suggest that a connection that has never been attempts should NOT be considered as failed. I agree with you on this. The current patch actually enforces this understanding for `connectionFailed(...)`, right?

Thus I think our concern is whether we should consider a connection as disconnected (i.e. `connectionStates.isDisconnected` returns `True`) or not when a connection has not even been attempted yet. Here is my thought:

- I prefer not to consider a connection as disconnected if no attempt has been made to connect to the node via e.g. `trySend()`. In this case, we can just use the existing patch.
- If we are not OK with the above definition, then we need a separate method `ClusterConnectionStates.connectionAttempted(...)`  to distinguish between states `connection not attempt` and `connection attempted and failed`. In this case, I will revert to use the first commit of this patch.
- As Jason mentioned, using results from `poll(...)` may not be sufficient.

@hachikuji Following your suggestion of adding `List<Node> disconnected()`, does this method return nodes which we haven't even attempted to connect to? If no, then we can probably just add `isDisconnected(...,)` to be consistent with existing methods `isReady(...)` and `isConnecting(...)`. If yes, then we still need a separate method such as `connectionAttempted(...)` to distinguish between states `connection not attempt` and `connection attempted and failed`.